tfhe/src/core_crypto/experimental/algorithms/test/lwe_stair_keyswitch.rs

```diff
@@ -0,0 +1,296 @@
+    partial_glwe_secret_key_fill: PartialGlweSecretKeyRandomCoefCount(2048),
+    bsk_glwe_noise_distribution: DynamicDistribution::new_gaussian_from_std_dev(StandardDev(
+        3.16202663074765e-16,
+    )),
+    lwe_dimension: LweDimension(664),
+    ks1_lwe_noise_distribution: DynamicDistribution::new_gaussian_from_std_dev(StandardDev(
+        7.60713313301797e-9,
+    )),
+    ks2_lwe_noise_distribution: DynamicDistribution::new_gaussian_from_std_dev(StandardDev(
+        0.0000380960250519291,
+    )),
+    pbs_level: DecompositionLevelCount(1),
+    pbs_base_log: DecompositionBaseLog(22),
+    ks1_level: DecompositionLevelCount(1),
+    ks1_base_log: DecompositionBaseLog(13),
+    ks2_level: DecompositionLevelCount(6),
+    ks2_base_log: DecompositionBaseLog(2),
+    ks1_unshared_coeff_count: LweSecretKeyUnsharedCoefCount(922),
+    ks2_unshared_coeff_count: LweSecretKeyUnsharedCoefCount(462),
+    ciphertext_modulus: CiphertextModulus::new_native(),
+};
+
+fn lwe_encrypt_stair_keyswitch_pbs_decrypt_custom_mod<
+    Scalar: UnsignedTorus + Send + Sync + CastFrom<usize> + CastInto<usize>,
+>(
+    params: StairKSParam<Scalar>,
+) {
+    let log_precision = params.log_precision;
+    let glwe_dimension = params.glwe_dimension;
+    let polynomial_size = params.polynomial_size;
+    let partial_glwe_secret_key_fill = params.partial_glwe_secret_key_fill;
+    let std_dev_bsk = params.bsk_glwe_noise_distribution;
+    let lwe_dimension = params.lwe_dimension;
+    let ks1_lwe_noise_distribution = params.ks1_lwe_noise_distribution;
+    let ks2_lwe_noise_distribution = params.ks2_lwe_noise_distribution;
+    let pbs_level = params.pbs_level;
+    let pbs_base_mpg = params.pbs_base_log;
+    let ks1_level = params.ks1_level;
+    let ks1_base_log = params.ks1_base_log;
+    let ks2_level = params.ks2_level;
+    let ks2_base_log = params.ks2_base_log;
+    let ks1_unshared_coeff_count = params.ks1_unshared_coeff_count;
+    let ks2_unshared_coeff_count = params.ks2_unshared_coeff_count;
+    let ciphertext_modulus = params.ciphertext_modulus;
+
+    let msg_modulus = Scalar::ONE.shl(log_precision.0);
+    let encoding_with_padding = get_encoding_with_padding(ciphertext_modulus);
+
+    let mut rsc = TestResources::new();
+
+    const NB_TESTS: usize = 10;
+    let mut msg = msg_modulus;
+    let delta: Scalar = encoding_with_padding / msg_modulus;
+
+    let large_lwe_dimension = glwe_dimension.to_equivalent_lwe_dimension(polynomial_size);
+    // Our algorithm is set up so that this equality holds, but in practice it could be more generic
+    // if we had an intermediate LweDimension being defined
+    assert_eq!(
+        large_lwe_dimension.0,
+        lwe_dimension.0 + ks1_unshared_coeff_count.0 + ks2_unshared_coeff_count.0
+    );
+
+    while msg != Scalar::ZERO {
+        msg = msg.wrapping_sub(Scalar::ONE);
+
+        let glwe_secret_key = allocate_and_generate_new_partial_binary_glwe_secret_key(
```

Comment:
Do we want to regenerate keys for each msg?

Reply:
Can't we simply allocate a simple glwe secre key here ?
I don't understand what `allocate_and_generate_new_partial_binary_glwe_secret_key` does exactly and why it's needed

Reply:
partial means some coefficients at the end are 0

Reply:
Why is it needed?
It would make more sense to me to generate a lwe sk and have a method to extend it to a glwe sk by adding zeros if necessary.